engine/flink/components/base-tests/src/test/scala/pl/touk/nussknacker/engine/flink/table/join/TableJoinTest.scala

```diff
@@ -0,0 +1,105 @@
+import pl.touk.nussknacker.engine.flink.table.FlinkTableComponentProvider
+import pl.touk.nussknacker.engine.flink.table.join.TableJoinTest.OrderProduct
+import pl.touk.nussknacker.engine.flink.test.FlinkSpec
+import pl.touk.nussknacker.engine.flink.util.transformer.join.BranchType
+import pl.touk.nussknacker.engine.util.test.TestScenarioRunner
+import pl.touk.nussknacker.test.ValidatedValuesDetailedMessage
+
+import scala.beans.BeanProperty
+
+class TableJoinTest extends AnyFunSuite with FlinkSpec with Matchers with Inside with ValidatedValuesDetailedMessage {
+
+  import pl.touk.nussknacker.engine.flink.util.test.FlinkTestScenarioRunner._
+  import pl.touk.nussknacker.engine.spel.SpelExtension._
+
+  import scala.jdk.CollectionConverters._
+
+  private lazy val additionalComponents: List[ComponentDefinition] =
+    FlinkTableComponentProvider.configIndependentComponents ::: Nil
+
+  private lazy val runner = TestScenarioRunner
+    .flinkBased(ConfigFactory.empty(), flinkMiniCluster)
+    .withExtraComponents(additionalComponents)
+    .build()
+
+  private val MainBranchId = "main"
+
+  private val JoinedBranchId = "joined"
+
+  private val JoinNodeId = "joined-node-id"
+
+  test("should be able to join") {
+    val scenario = ScenarioBuilder
+      .streaming("sample-join-last")
+      .sources(
+        GraphBuilder
+          .source("orders-source", TestScenarioRunner.testDataSource)
+          .filter("orders-filter", "#input.type == 'order'".spel)
+          .branchEnd(MainBranchId, JoinNodeId),
+        GraphBuilder
+          .source("products-source", TestScenarioRunner.testDataSource)
+          .filter("product-filter", "#input.type == 'product'".spel)
+          .branchEnd(JoinedBranchId, JoinNodeId),
+        GraphBuilder
+          .join(
+            JoinNodeId,
+            "join",
+            Some("product"),
+            List(
+              MainBranchId -> List(
+                "branchType" -> s"T(${classOf[BranchType].getName}).MAIN".spel,
+                "key"        -> s"#input.productId.toString".spel
+              ),
+              JoinedBranchId -> List(
+                "branchType" -> s"T(${classOf[BranchType].getName}).JOINED".spel,
+                "key"        -> s"#input.id.toString".spel
+              )
+            ),
+            "output" -> "#input".spel,
+          )
+          .emptySink("end", TestScenarioRunner.testResultSink, "value" -> "{#input, #product}".spel)
+      )
+
+    val result = runner.runWithData(
+      scenario,
+      List(
+        OrderProduct("product", 1, -1),
+        OrderProduct("order", 10, 1),
+      ),
+      Boundedness.BOUNDED,
+      Some(RuntimeExecutionMode.BATCH)
+    )
+
+    result.validValue.successes shouldBe List(
+      List(OrderProduct("order", 10, 1), OrderProduct("product", 1, -1)).asJava,
+    )
+  }
+
+}
+
+object TableJoinTest {
+
+  // TODO: split into separate classes and pass two streams to separate source nodes
```

Comment:
Yes, I'd split e.g. to Order, Product, and OrderWithProduct.
At first glance I thought that you implemented union instead of join (the nested list in the output is not so visible)

Reply:
So far, I only polished a bit current solution (more meaningful class name, factory methods). I also changed returned type to record - take a look. Splitting it into two separate classes would be expensive - we have to add some branch-aware methods to scenario runners, so I didn't do it.